src/server/pps/server/api_server.go

```diff
@@ -1377,127 +1385,37 @@ func (a *apiServer) GetLogs(request *pps.GetLogsRequest, apiGetLogsServer pps.AP
-	// (sort the pods to make sure that the order of log lines is stable)
-	sort.Sort(podSlice(pods))
-	logCh := make(chan *pps.LogMessage)
-	var eg errgroup.Group
-	var mu sync.Mutex
-	eg.Go(func() error {
-		for _, pod := range pods {
-			pod := pod
-			if !request.Follow {
-				mu.Lock()
+		return lokiutil.ForEachLine(resp, func(line string) error {
+			msg := &pps.LogMessage{}
+			if err := jsonpb.Unmarshal(strings.NewReader(line), msg); err != nil {
+				return nil
 			}
-			eg.Go(func() (retErr error) {
-				if !request.Follow {
-					defer mu.Unlock()
-				}
-				tailLines := &request.Tail
-				if *tailLines <= 0 {
-					tailLines = nil
-				}
-				// Get full set of logs from pod i
-				stream, err := a.env.GetKubeClient().CoreV1().Pods(a.namespace).GetLogs(
-					pod.ObjectMeta.Name, &v1.PodLogOptions{
-						Container: containerName,
-						Follow:    request.Follow,
-						TailLines: tailLines,
-					}).Timeout(10 * time.Second).Stream()
-				if err != nil {
-					return err
-				}
-				defer func() {
-					if err := stream.Close(); err != nil && retErr == nil {
-						retErr = err
-					}
-				}()
-
-				// Parse pods' log lines, and filter out irrelevant ones
-				scanner := bufio.NewScanner(stream)
-				for scanner.Scan() {
-					msg := new(pps.LogMessage)
-					if containerName == "pachd" {
-						msg.Message = scanner.Text()
-					} else {
-						logBytes := scanner.Bytes()
-						if err := jsonpb.Unmarshal(bytes.NewReader(logBytes), msg); err != nil {
-							continue
-						}
-
-						// Filter out log lines that don't match on pipeline or job
-						if request.Pipeline != nil && request.Pipeline.Name != msg.PipelineName {
-							continue
-						}
-						if request.Job != nil && request.Job.ID != msg.JobID {
-							continue
-						}
-						if request.Datum != nil && request.Datum.ID != msg.DatumID {
-							continue
-						}
-						if request.Master != msg.Master {
-							continue
-						}
-						if !workercommon.MatchDatum(request.DataFilters, msg.Data) {
-							continue
-						}
-					}
-					msg.Message = strings.TrimSuffix(msg.Message, "\n")
 
-					// Log message passes all filters -- return it
-					select {
-					case logCh <- msg:
-					case <-ctx.Done():
-						return nil
-					}
-				}
+			// Filter out log lines that don't match on pipeline or job
+			if request.Pipeline != nil && request.Pipeline.Name != msg.PipelineName {
 				return nil
-			})
-		}
-		return nil
-	})
-	var egErr error
-	go func() {
-		egErr = eg.Wait()
-		close(logCh)
-	}()
-
-	for msg := range logCh {
-		if err := apiGetLogsServer.Send(msg); err != nil {
-			return err
-		}
+			}
+			if request.Job != nil && request.Job.ID != msg.JobID {
+				return nil
+			}
+			if request.Datum != nil && request.Datum.ID != msg.DatumID {
+				return nil
+			}
+			if request.Master != msg.Master {
+				return nil
+			}
+			if !workercommon.MatchDatum(request.DataFilters, msg.Data) {
+				return nil
+			}
```

Comment:
could these filters be applied to the loki query instead?

Reply:
Updated the loki query to contain these filters, they're still implemented in our code as well because some of the loki queries are just string matching, so technically users can log something that matches the filters and inject spurious log lines into the results. That should be pretty rare, but doing it on both places means that if users do that all that happens is performance gets a little worse since we stream messages back from loki but then filter them out client side.